File: src/map_app/static/map_app/js/map_search_contract.js
(function(global) {
    var MARKER_FIELDS = [
        "activity_log_id",
        "performance_id",
        "location_id",
        "location_name",
        "date",
        "lat",
        "lng",
        "icon_color",
    ];

    var SUMMARY_FIELDS = [
        "total_locations",
        "total_activity_logs",
        "total_performances",
        "marker_count",
        "tagged_locations",
        "new_count",
        "revisit_count",
    ];
    var STATISTICS_FIELDS = [
        "month_labels",
        "month_values",
        "recent_visits",
        "top_locations",
    ];
    var RECENT_VISIT_FIELDS = [
        "location_name",
        "date",
        "title",
        "latitude",
        "longitude",
    ];
    var TOP_LOCATION_FIELDS = [
        "name",
        "count",
        "latitude",
        "longitude",
    ];

    var MAP_SEARCH_RESPONSE_SCHEMA = {
        type: "object",
        required: ["markers", "summary", "statistics"],
        additionalProperties: false,
        properties: {
            markers: {
                type: "array",
                items: {
                    type: "object",
                    required: MARKER_FIELDS.slice(),
                    additionalProperties: false,
                    properties: {
                        activity_log_id: { type: "integer" },
                        performance_id: { type: "integer" },
                        location_id: { type: "integer" },
                        location_name: { type: "string" },
                        date: { type: "string" },
                        lat: { type: "number" },
                        lng: { type: "number" },
                        icon_color: { type: "string" },
                    },
                },
            },
            summary: {
                type: "object",
                required: SUMMARY_FIELDS.slice(),
                additionalProperties: false,
                properties: {
                    total_locations: { type: "integer" },
                    total_activity_logs: { type: "integer" },
                    total_performances: { type: "integer" },
                    marker_count: { type: "integer" },
                    tagged_locations: { type: "integer" },
                    new_count: { type: "integer" },
                    revisit_count: { type: "integer" },
                },
            },
            statistics: {
                type: "object",
                required: STATISTICS_FIELDS.slice(),
                additionalProperties: false,
                properties: {
                    month_labels: { type: "array", items: { type: "string" } },
                    month_values: { type: "array", items: { type: "integer" } },
                    recent_visits: {
                        type: "array",
                        items: {
                            type: "object",
                            required: RECENT_VISIT_FIELDS.slice(),
                            additionalProperties: false,
                            properties: {
                                location_name: { type: "string" },
                                date: { type: "string" },
                                title: { type: "string" },
                                latitude: { type: "number" },
                                longitude: { type: "number" },
                            },
                        },
                    },
                    top_locations: {
                        type: "array",
                        items: {
                            type: "object",
                            required: TOP_LOCATION_FIELDS.slice(),
                            additionalProperties: false,
                            properties: {
                                name: { type: "string" },
                                count: { type: "integer" },
                                latitude: { type: "number" },
                                longitude: { type: "number" },
                            },
                        },
                    },
                },
            },
        },
    };

    function hasExactlyKeys(target, expectedKeys) {
        var keys = Object.keys(target).sort();
        var sortedExpected = expectedKeys.slice().sort();
        if (keys.length !== sortedExpected.length) return false;
        for (var i = 0; i < keys.length; i += 1) {
            if (keys[i] !== sortedExpected[i]) return false;
        }
        return true;
    }

    function assertInteger(value, fieldName) {
        if (!Number.isInteger(value)) {
            throw new Error("Invalid map search payload: '" + fieldName + "' must be integer.");
        }
    }

    function assertFiniteNumber(value, fieldName) {
        if (typeof value !== "number" || !Number.isFinite(value)) {
            throw new Error("Invalid map search payload: '" + fieldName + "' must be number.");
        }
    }

    function assertString(value, fieldName) {
        if (typeof value !== "string") {
            throw new Error("Invalid map search payload: '" + fieldName + "' must be string.");
        }
    }

    function validateMapSearchPayload(payload) {
        if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
            throw new Error("Invalid map search payload: root must be object.");
        }
        if (!hasExactlyKeys(payload, MAP_SEARCH_RESPONSE_SCHEMA.required)) {
            throw new Error("Invalid map search payload: root keys mismatch.");
        }
        if (!Array.isArray(payload.markers)) {
            throw new Error("Invalid map search payload: 'markers' must be array.");
        }
        payload.markers.forEach(function(marker, index) {
            if (!marker || typeof marker !== "object" || Array.isArray(marker)) {
                throw new Error("Invalid map search payload: marker[" + index + "] must be object.");
            }
            if (!hasExactlyKeys(marker, MARKER_FIELDS)) {
                throw new Error("Invalid map search payload: marker keys mismatch.");
            }
            assertInteger(marker.activity_log_id, "activity_log_id");
            assertInteger(marker.performance_id, "performance_id");
            assertInteger(marker.location_id, "location_id");
            assertString(marker.location_name, "location_name");
            assertString(marker.date, "date");
            assertFiniteNumber(marker.lat, "lat");
            assertFiniteNumber(marker.lng, "lng");
            assertString(marker.icon_color, "icon_color");
        });

        var summary = payload.summary;
        if (!summary || typeof summary !== "object" || Array.isArray(summary)) {
            throw new Error("Invalid map search payload: 'summary' must be object.");
        }
        if (!hasExactlyKeys(summary, SUMMARY_FIELDS)) {
            throw new Error("Invalid map search payload: summary keys mismatch.");
        }
        assertInteger(summary.total_locations, "total_locations");
        assertInteger(summary.total_activity_logs, "total_activity_logs");
        assertInteger(summary.total_performances, "total_performances");
        assertInteger(summary.marker_count, "marker_count");
        assertInteger(summary.tagged_locations, "tagged_locations");
        assertInteger(summary.new_count, "new_count");
        assertInteger(summary.revisit_count, "revisit_count");

        var statistics = payload.statistics;
        if (!statistics || typeof statistics !== "object" || Array.isArray(statistics)) {
            throw new Error("Invalid map search payload: 'statistics' must be object.");
        }
        if (!hasExactlyKeys(statistics, STATISTICS_FIELDS)) {
            throw new Error("Invalid map search payload: statistics keys mismatch.");
        }

        if (!Array.isArray(statistics.month_labels)) {
            throw new Error("Invalid map search payload: 'month_labels' must be array.");
        }
        statistics.month_labels.forEach(function(label) {
            assertString(label, "month_labels");
        });

        if (!Array.isArray(statistics.month_values)) {
            throw new Error("Invalid map search payload: 'month_values' must be array.");
        }
        statistics.month_values.forEach(function(value) {
            assertInteger(value, "month_values");
        });

        if (!Array.isArray(statistics.recent_visits)) {
            throw new Error("Invalid map search payload: 'recent_visits' must be array.");
        }
        statistics.recent_visits.forEach(function(visit, index) {
            if (!visit || typeof visit !== "object" || Array.isArray(visit)) {
                throw new Error("Invalid map search payload: recent_visits[" + index + "] must be object.");
            }
            if (!hasExactlyKeys(visit, RECENT_VISIT_FIELDS)) {
                throw new Error("Invalid map search payload: recent visit keys mismatch.");
            }
            assertString(visit.location_name, "location_name");
            assertString(visit.date, "date");
            assertString(visit.title, "title");
            assertFiniteNumber(visit.latitude, "latitude");
            assertFiniteNumber(visit.longitude, "longitude");
        });

        if (!Array.isArray(statistics.top_locations)) {
            throw new Error("Invalid map search payload: 'top_locations' must be array.");
        }
        statistics.top_locations.forEach(function(item, index) {
            if (!item || typeof item !== "object" || Array.isArray(item)) {
                throw new Error("Invalid map search payload: top_locations[" + index + "] must be object.");
            }
            if (!hasExactlyKeys(item, TOP_LOCATION_FIELDS)) {
                throw new Error("Invalid map search payload: top location keys mismatch.");
            }
            assertString(item.name, "name");
            assertInteger(item.count, "count");
            assertFiniteNumber(item.latitude, "latitude");
            assertFiniteNumber(item.longitude, "longitude");
        });
    }

    global.MapSearchContract = {
        MAP_SEARCH_RESPONSE_SCHEMA: MAP_SEARCH_RESPONSE_SCHEMA,
        MARKER_FIELDS: MARKER_FIELDS,
        SUMMARY_FIELDS: SUMMARY_FIELDS,
        validateMapSearchPayload: validateMapSearchPayload,
    };
})(window);

File: src/map_app/contracts/map_search_contract.py
from dataclasses import asdict, dataclass
from typing import Any


@dataclass(frozen=True)
class MapSearchSummary:
    total_locations: int
    total_activity_logs: int
    total_performances: int
    marker_count: int
    tagged_locations: int
    new_count: int
    revisit_count: int


@dataclass(frozen=True)
class MapSearchMarker:
    activity_log_id: int
    performance_id: int
    location_id: int
    location_name: str
    date: str
    lat: float
    lng: float
    icon_color: str


@dataclass(frozen=True)
class MapSearchRecentVisit:
    location_name: str
    date: str
    title: str
    latitude: float
    longitude: float


@dataclass(frozen=True)
class MapSearchTopLocation:
    name: str
    count: int
    latitude: float
    longitude: float


@dataclass(frozen=True)
class MapSearchStatistics:
    month_labels: list[str]
    month_values: list[int]
    recent_visits: list[MapSearchRecentVisit]
    top_locations: list[MapSearchTopLocation]


@dataclass(frozen=True)
class MapSearchPayload:
    markers: list[MapSearchMarker]
    summary: MapSearchSummary
    statistics: MapSearchStatistics

    def to_dict(self) -> dict[str, Any]:
        return {
            "markers": [asdict(marker) for marker in self.markers],
            "summary": asdict(self.summary),
            "statistics": asdict(self.statistics),
        }


MAP_SEARCH_SUMMARY_FIELDS = tuple(MapSearchSummary.__dataclass_fields__.keys())
MAP_SEARCH_MARKER_FIELDS = tuple(MapSearchMarker.__dataclass_fields__.keys())
MAP_SEARCH_STATISTICS_FIELDS = tuple(MapSearchStatistics.__dataclass_fields__.keys())
MAP_SEARCH_RECENT_VISIT_FIELDS = tuple(MapSearchRecentVisit.__dataclass_fields__.keys())
MAP_SEARCH_TOP_LOCATION_FIELDS = tuple(MapSearchTopLocation.__dataclass_fields__.keys())

MAP_SEARCH_RESPONSE_SCHEMA = {
    "type": "object",
    "required": ["markers", "summary", "statistics"],
    "additionalProperties": False,
    "properties": {
        "markers": {
            "type": "array",
            "items": {
                "type": "object",
                "required": list(MAP_SEARCH_MARKER_FIELDS),
                "additionalProperties": False,
                "properties": {
                    "activity_log_id": {"type": "integer"},
                    "performance_id": {"type": "integer"},
                    "location_id": {"type": "integer"},
                    "location_name": {"type": "string"},
                    "date": {"type": "string"},
                    "lat": {"type": "number"},
                    "lng": {"type": "number"},
                    "icon_color": {"type": "string"},
                },
            },
        },
        "summary": {
            "type": "object",
            "required": list(MAP_SEARCH_SUMMARY_FIELDS),
            "additionalProperties": False,
                "properties": {
                    "total_locations": {"type": "integer"},
                    "total_activity_logs": {"type": "integer"},
                    "total_performances": {"type": "integer"},
                "marker_count": {"type": "integer"},
                "tagged_locations": {"type": "integer"},
                "new_count": {"type": "integer"},
                "revisit_count": {"type": "integer"},
            },
        },
        "statistics": {
            "type": "object",
            "required": list(MAP_SEARCH_STATISTICS_FIELDS),
            "additionalProperties": False,
            "properties": {
                "month_labels": {"type": "array", "items": {"type": "string"}},
                "month_values": {"type": "array", "items": {"type": "integer"}},
                "recent_visits": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "required": list(MAP_SEARCH_RECENT_VISIT_FIELDS),
                        "additionalProperties": False,
                        "properties": {
                            "location_name": {"type": "string"},
                            "date": {"type": "string"},
                            "title": {"type": "string"},
                            "latitude": {"type": "number"},
                            "longitude": {"type": "number"},
                        },
                    },
                },
                "top_locations": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "required": list(MAP_SEARCH_TOP_LOCATION_FIELDS),
                        "additionalProperties": False,
                        "properties": {
                            "name": {"type": "string"},
                            "count": {"type": "integer"},
                            "latitude": {"type": "number"},
                            "longitude": {"type": "number"},
                        },
                    },
                },
            },
        },
    },
}


def validate_map_search_payload(payload: dict[str, Any]) -> None:
    if not isinstance(payload, dict):
        raise ValueError("Map search payload must be a JSON object.")
